src/Datadog.Trace/Agent/MovingAverageKeepRateCalculator.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Agent
+{
+    /// <summary>
+    /// Used to calculate the Trace Keep Rate, tracking the number of
+    /// traces kept and dropped that should have been sent to the agent.
+    /// Traces that are subsequently dropped by the agent due to sampling
+    /// will not count as dropped in this rate.
+    /// </summary>
+    internal class MovingAverageKeepRateCalculator : IKeepRateCalculator
+    {
+        private const int DefaultWindowSize = 10;
+        private static readonly TimeSpan DefaultBucketDuration = TimeSpan.FromSeconds(1);
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<MovingAverageKeepRateCalculator>();
+
+        private readonly int _windowSize;
+        private readonly TimeSpan _bucketDuration;
+        private readonly uint[] _dropped;
+        private readonly uint[] _created;
+
+        private readonly CancellationTokenSource _cts = new();
+        private readonly TaskCompletionSource<bool> _processExit = new TaskCompletionSource<bool>();
+
+        private int _index = 0;
+        private long _totals = 0;
+
+        private long _latestDrops = 0;
+        private long _latestKeeps = 0;
+
+        internal MovingAverageKeepRateCalculator(int windowSize, TimeSpan bucketDuration)
+        {
+            if (windowSize < 0 || windowSize > 100)
+            {
+                throw new ArgumentOutOfRangeException(nameof(windowSize), windowSize, "Must be a value between 1 and 100");
+            }
+
+            _windowSize = windowSize;
+            _bucketDuration = bucketDuration;
+            _dropped = new uint[windowSize];
+            _created = new uint[windowSize];
+
+            if (bucketDuration != Timeout.InfiniteTimeSpan)
+            {
+                Task.Run(UpdateBucketTaskLoopAsync)
+                    .ContinueWith(t => Log.Error(t.Exception, "Error in "), TaskContinuationOptions.OnlyOnFaulted);
+            }
+        }
+
+        public static MovingAverageKeepRateCalculator CreateDefaultKeepRateCalculator()
+            => new MovingAverageKeepRateCalculator(DefaultWindowSize, DefaultBucketDuration);
+
+        /// <summary>
+        /// Increment the number of kept traces
+        /// </summary>
+        public void IncrementKeeps(int count)
+        {
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
```

Comment:
Can we move this to a ThrowHelper static class/method, I know there is a lot of code like this, but this method is a good candidate for inlining due the size. But for now the JIT is no able to inline methods that throws or that have try/catch blocks.

Reply:
Thinking about that, I will make a new PR to catch all the throws with a global `ThrowHelper` static class.